### DISCUSSION TITLE
setFormTypeOptions

### DISCUSSION THREAD

**Discussion by romanpr76** [symfony/symfony#44041] [Symfony Book: The Fast Track]
Hi! There is a code samle in a book like this:

```php
$createdAt = DateTimeField::new('createdAt')->setFormTypeOptions([
    'html5' => true,
    'years' => range(date('Y'), date('Y') + 5),
    'widget' => 'singleText',
]);
```
where can I find more information about these options? The list of available options for each field type and the meaning of them.

**Comment by BahmanMehrdad** [comment]
More information can be found at below link 

[Form Types Reference](https://symfony.com/doc/current/reference/forms/types.html)

**Reply by romanpr76** [reply]
Thank you, but I'm confused a little.

There were told

_widget
Defines the widget option for both the DateType and TimeType. This can be overridden with the date_widget and time_widget options._

But there is a DateTimeField in the example from the book. Is it a mistake?
Ok, I've changet it to DateField

There were also told:

_years
type: array default: five years before to five years after the current year
List of years available to the year field type. This option is only relevant when the widget option is set to choice._

Ok, I've changed **widget** also to **choice**

And now I have this:

```
$createdAt = DateField::new('createdAt')->setFormTypeOptions([
  'html5' => true,
  'widget' => 'choice',
  'years' => range(date('Y'), date('Y') + 5),
]);
```

But nothing is changed. Looks like it does'n have any effect. What did I wrong?

**Comment by stof** [comment]
This code snippet in the book is in the section `Customizing EasyAdmin` of the chapter `Setting up an Admin Backend` which shows the usage of easy-admin-bundle.

The actual code in the book also includes a use statement `use EasyCorp\Bundle\EasyAdminBundle\Field\DateTimeField;`, which makes it clear that this API belongs to EasyAdminBundle, not to Symfony. The doc of EasyAdminBundle is available on https://symfony.com/doc/3.x/EasyAdminBundle/index.html
In that case, `setFormTypeOptions` is a method allowing to set the options for the underlying form type. In the case of `DateTimeField`, the underlying type is a `DateTimeType` of Symfony, for which the available options are documented in the form type reference of the Symfony docs

**Reply by romanpr76** [reply]
Thank you too, but I still can't find any info about setFormTypeOptions and its options. In the [(EasyAdminBundle docs)](https://symfony.com/doc/3.x/EasyAdminBundle/fields.html) I've fond only a list of avaliable types and nothing more.
I've tried to change some options, but nothing changed. I have no idea what does this code in the book do. I have the same result if I just completely remove the call of setFormTypeOptions.

**Reply by stof** [reply]
I suggest you to ask this t the EasyAdminBundle community, as you are more likely to get an answer (I never used that bundle myself). They have a dedicated channel on the Symfony slack.

**Reply by romanpr76** [reply]
Ok, thanx!
